frontend/src/utils/generate-warnings.ts

```diff
@@ -75,6 +90,473 @@ export function produceWarnings(schedule: Schedule): IWarning[] {
+ * @param creditHoursNeeded num hours need to satisfy top-level requirement.
+ */
+function processIAndCourse(
+  requirement: IAndCourse,
+  taken: Map<string, number>,
+  satisfied: CreditHourTracker,
+  coursesUsed: Set<string>,
+  creditHoursNeeded: number
+): string | undefined {
+  // requirement is unsatisfied if even one of the requirements in the list is unsatisfied.
+  let processReqMessages: string[] = processRequirementCourses(
+    requirement.courses,
+    taken,
+    satisfied,
+    coursesUsed,
+    creditHoursNeeded
+  );
+  if (processReqMessages.length > 0) {
+    let reqMessage = `(${processReqMessages.join(" and ")})`;
+    return reqMessage;
+  } else {
+    return undefined;
+  }
+}
+
+/**
+ * Processes an IOrCourse requirement.
+ * @param requirement the requirement to check
+ * @param taken the Map of courses a student has on their schedule right now
+ * @param satisfied a tracker of the hours satisfied for the top-level requirement.
+ * @param coursesUsed a set of courses which already satisfy some requirement for the major.
+ * @param creditHoursNeeded num hours need to satisfy top-level requirement.
+ */
+function processIOrCourse(
+  requirement: IOrCourse,
+  taken: Map<string, number>,
+  satisfied: CreditHourTracker,
+  coursesUsed: Set<string>,
+  creditHoursNeeded: number
+): string | undefined {
+  // requirement is unsatisfied if all of the requirements in the list have not been satisfied.
+  let processReqMessages: string[] = processRequirementCourses(
+    requirement.courses,
+    taken,
+    satisfied,
+    coursesUsed,
+    creditHoursNeeded
+  );
+  if (processReqMessages.length === requirement.courses.length) {
+    let reqMessage = `(${processReqMessages.join(" or ")})`;
+    return reqMessage;
+  } else {
+    return undefined;
+  }
+}
+
+/**
+ * Processes an ICourseRange requirement.
+ * @param requirement the requirement to check
+ * @param taken the Map of courses a student has on their schedule right now
+ * @param satisfied a tracker of the hours satisfied for the top-level requirement.
+ * @param coursesUsed a set of courses which already satisfy some requirement for the major.
+ * @param creditHoursNeeded num hours need to satisfy top-level requirement.
+ */
+function processICourseRange(
+  requirement: ICourseRange,
+  taken: Map<string, number>,
+  satisfied: CreditHourTracker,
+  coursesUsed: Set<string>,
+  creditHoursNeeded: number
+): string | undefined {
+  // requirement is unsatisfied if the numCredits for the courseRange hasn't been satisfied.
+  let numCreditsrequired: number = requirement.creditsRequired;
+  let rangeCreditsCompleted: number = 0;
+
+  //loop through the taken courses and check if it is in one of the subject ranges for this ICourseRange.
+  taken.forEach((value: number, key: string) => {
+    //check the global map to see if it has not already been used.
+    if (!coursesUsed.has(key)) {
+      if (courseInSubjectRanges(key, requirement.ranges)) {
```

Comment:
It would be slightly more memory intensive, but here it makes more sense to simply store a set of ranges, in an object. Then you could avoid having to split the courseCode, which is bad. Alternatively, you could store courses not as a course code, but as an object containing just a subject and classId.